### PR TITLE
Require C++17

### PR DESCRIPTION
**Pull request by pdillinger**
Summary: Drop support for some old compilers by requiring C++17 standard
(or higher). See #9388

First modification based on this is to remove some conditional compilation in slice.h (also
better for ODR)

Also in this PR:
* Fix some Makefile formatting that seems to affect ASSERT_STATUS_CHECKED config in
some cases
* Add c_test to NON_PARALLEL_TEST in Makefile
* Fix a clang-analyze reported "potential leak" in lru_cache_test
* Better "compatibility" definition of DEFINE_uint32 for old versions of gflags
* Fix a linking problem with shared libraries in Makefile (`./random_test: error while loading shared libraries: librocksdb.so.6.29: cannot open shared object file: No such file or directory`)
* Always set ROCKSDB_SUPPORT_THREAD_LOCAL and use thread_local (from C++11)
  * TODO in later PR: clean up that obsolete flag
* Fix a cosmetic typo in c.h (#9488)

Test Plan: CircleCI config substantially updated.

* Upgrade to latest Ubuntu images for each release
* Generally prefer Ubuntu 20, but keep a couple Ubuntu 16 builds with oldest supported
compilers, to ensure compatibility
* Remove .circleci/cat_ignore_eagain except for Ubuntu 16 builds, because this is to work
around a kernel bug that should not affect anything but Ubuntu 16.
* Remove designated gcc-9 build, because the default linux build now uses GCC 9 from
Ubuntu 20.
* Add some `apt-key add` to fix some apt "couldn't be verified" errors
* Generally drop SKIP_LINK=1; work-around no longer needed
* Generally `add-apt-repository` before `apt-get update` as manual testing indicated the
reverse might not work.

Travis:
* Use gcc-7 by default (remove specific gcc-7 and gcc-4.8 builds)
* TODO in later PR: fix s390x "Assembler messages: Error: invalid switch -march=z14" failure

AppVeyor:
* Completely dropped because we are dropping VS2015 support and CircleCI covers
VS >= 2017

Also local testing with old gflags (out of necessity when using ROCKSDB_NO_FBCODE=1).

### CODE REVIEW

**Comment by pdillinger** [comment]
> I wonder if using "-j$(nproc)" might be a better paradigm

I'm not touching this at this time.

> What happens if the examples do not specify a --std? Won't it build with the default for the compiler? And if that is less than 17, what will happen?

You'll get errors like undefined `std::string_view`

File: CMakeLists.txt [hunk 1]
```diff
@@ -369,18 +369,8 @@ endif()
...
-#endif
-int main() {
-  static __thread int tls;
-  (void)tls;
-}
-" HAVE_THREAD_LOCAL)
-if(HAVE_THREAD_LOCAL)
-  add_definitions(-DROCKSDB_SUPPORT_THREAD_LOCAL)
-endif()
+# Thread-local is part of C++11 (TODO: clean up this flag)
```
Comment (mrambacher) [inline CMakeLists.txt:372]: C++17 or part of the C++ standard.  (It is a bit confusing to say C++11 when we say we support 17 only.

File: INSTALL.md [hunk 2]
```diff
@@ -50,7 +50,7 @@ to build a portable binary, add `PORTABLE=1` before your make commands, like thi
 ## Supported platforms
 
 * **Linux - Ubuntu**
-    * Upgrade your gcc to version at least 4.8 to get C++11 support.
+    * Upgrade your gcc to version at least 7 to get C++17 support.
```
Comment (mrambacher) [inline INSTALL.md:53]: Should this mention clang?

File: include/rocksdb/slice.h [hunk 3]
```diff
@@ -44,11 +41,9 @@ class Slice {
   /* implicit */
   Slice(const std::string& s) : data_(s.data()), size_(s.size()) {}
 
-#ifdef __cpp_lib_string_view
   // Create a slice that refers to the same contents as "sv"
   /* implicit */
   Slice(std::string_view sv) : data_(sv.data()), size_(sv.size()) {}
```
Comment (mrambacher) [inline include/rocksdb/slice.h:46]: Should this pass in a const std::string_view& ?

File: .circleci/config.yml [hunk 4]
```diff
@@ -181,153 +183,153 @@ jobs:
...
     machine:
-      image: ubuntu-1604:202104-01
+      image: ubuntu-2004:202111-02
     resource_class: 2xlarge
     steps:
       - pre-steps
       - install-gflags
       - upgrade-cmake
-      - run: (mkdir build && cd build && cmake -DWITH_GFLAGS=1 .. && make V=1 -j20 && ctest -j20) 2>&1 | .circleci/cat_ignore_eagain
+      - run: (mkdir build && cd build && cmake -DWITH_GFLAGS=1 .. && make V=1 -j20 && ctest -j20)
```
[the remark this answers is not in the document]
Comment (pdillinger) [reply .circleci/config.yml:332]: This PR is not about parallelism in CircleCI

File: .circleci/config.yml [hunk 5]
```diff
@@ -338,7 +340,7 @@ jobs:
       - pre-steps
       - install-gflags
       - install-benchmark
-      - run: (mkdir build && cd build && cmake -DWITH_GFLAGS=1 -DWITH_BENCHMARK=1 .. && make V=1 -j20 && ctest -j20) 2>&1 | .circleci/cat_ignore_eagain
+      - run: (mkdir build && cd build && cmake -DWITH_GFLAGS=1 -DWITH_BENCHMARK=1 .. && make V=1 -j20 && ctest -j20)
```
[the remark this answers is not in the document]
Comment (pdillinger) [reply .circleci/config.yml:343]: This PR is not about parallelism in CircleCI

File: .circleci/config.yml [hunk 6]
```diff
@@ -348,79 +350,69 @@ jobs:
...
   build-linux-microbench:
     machine:
-      image: ubuntu-2004:202010-01
+      image: ubuntu-2004:202111-02
     resource_class: xlarge
     steps:
       - pre-steps
       - install-benchmark
-      - run: DEBUG_LEVEL=0 make microbench 2>&1 | .circleci/cat_ignore_eagain
+      - run: DEBUG_LEVEL=0 make microbench
```
[the remark this answers is not in the document]
Comment (pdillinger) [reply .circleci/config.yml:415]: This PR is not about parallelism in CircleCI

File: .circleci/config.yml [hunk 7]
```diff
@@ -608,25 +600,25 @@ jobs:
...
+      image: ubuntu-2004:202111-02
     resource_class: large
     steps:
       - pre-steps
       - install-gflags
       - run:
           name: "Build examples"
           command: |
-            OPT=-DTRAVIS V=1 make -j4 static_lib && cd examples && make -j4 | ../.circleci/cat_ignore_eagain
+            OPT=-DTRAVIS V=1 make -j4 static_lib && cd examples && make -j4
```
[the remark this answers is not in the document]
Comment (pdillinger) [reply .circleci/config.yml:616]: This PR is not about parallelism in CircleCI

File: INSTALL.md (hunk 2 above)
Comment (pdillinger) [reply INSTALL.md:53]: It's not the default compiler on Ubuntu

Comment (mrambacher) [reply INSTALL.md:53]: Understood that it is not the default compiler.  Nowhere in INSTALL does it say we support clang (on any platform) and what the minimum clang version is.  Can we add that somewhere?

File: CMakeLists.txt (hunk 1 above)
Comment (pdillinger) [reply CMakeLists.txt:372]: C++11 is factually accurate. I'm pointing out that the check has been effectively obsolete for a long time.

File: include/rocksdb/slice.h (hunk 3 above)
Comment (pdillinger) [reply include/rocksdb/slice.h:46]: OK

File: INSTALL.md (hunk 2 above)
Comment (pdillinger) [reply INSTALL.md:53]: It's near the top, "We do depend on newer gcc/clang with C++17 support (GCC >= 7, Clang >= 5)."

**Pull request closed without merging**